src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala

```diff
@@ -355,6 +355,7 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
                     && !sym.owner.isTrait
                     && sym.owner.enclosingPackageClass != currentClass.enclosingPackageClass
                     && qual.symbol.info.member(sym.name).exists
+                    && !(currentClass.typeOfThis <:< sym.owner.tpe) // scala/bug#11924
```

Comment:
Hmm, this is not enough for the REPL. Imitating the REPL, the following still prints `BA` instead of recursing:

```scala
package l1 {
class I {
  class A {
    protected def f(): Unit = print("A")
  }
}
object O extends I
}

package l2 {
class I {
  trait B { self: l1.O.A =>
    override def f(): Unit = {
      print("B")
      self.f()
    }
  }
}
object O extends I
}

class C extends l1.O.A with l2.O.B
object Test {
  def main(args: Array[String]): Unit = {
    new C().f()
  }
}
```

Reply:
Fixed.